File: app/assets/javascripts/AppMain.js
(function() {

    /**
     * ShockTrade Angular.js Application
     */
    var app = angular.module('shocktrade', ['ngAnimate', 'ngCookies', 'ngRoute', 'ngSanitize', 'nvd3ChartDirectives', 'toaster', 'ui.bootstrap']);

    /**
     * Configure the application
     */
    app.config(['$routeProvider',
        function ($routeProvider) {
            // setup the routes
            $routeProvider
                .when('/dashboard', {templateUrl: '/assets/views/dashboard/dashboard.htm', controller: 'DashboardController'})
                .when('/dashboard/:contestId', {templateUrl: '/assets/views/dashboard/dashboard.htm', controller: 'DashboardController'})
                .when('/discover', {templateUrl: '/assets/views/discover/discover.htm', controller: 'DiscoverController'})
                .when('/discover/:symbol', {templateUrl: '/assets/views/discover/discover.htm', controller: 'DiscoverController'})
                .when('/explore', {templateUrl: '/assets/views/explore/drill_down.htm', controller: 'DrillDownController'})
                .when('/inspect/:contestId', {templateUrl: '/assets/views/admin/inspect.htm', controller: 'InspectController'})
                .when('/news', {templateUrl: '/assets/views/news/news_center.htm', controller: 'NewsController'})
                .when('/research', {templateUrl: '/assets/views/research/research.htm', controller: 'ResearchController'})
                .when('/search', {templateUrl: '/assets/views/play/search.htm', controller: 'GameSearchController'})
                .when('/profile', {redirectTo: '/profile/connect'})
                .when('/profile/awards', {templateUrl: '/assets/views/profile/awards.htm', controller: 'ProfileController'})
                .when('/profile/connect', {templateUrl: '/assets/views/profile/connect/connect.htm', controller: 'ProfileController'})
                .when('/profile/statistics', {templateUrl: '/assets/views/profile/statistics.htm', controller: 'ProfileController'})
                .otherwise({redirectTo: '/discover'});
        }]);

    /**
     * Initialize the application
     */
    app.run(function ($rootScope,
                      $location,
                      Facebook,
                      FavoriteSymbols,
                      HeldSecurities,
                      MarketStatus,
                      MySession,
                      NewsSymbols,
                      QuoteService,
                      RecentSymbols,
                      WebSockets) {
        $rootScope.Facebook = Facebook;
        $rootScope.FavoriteSymbols = FavoriteSymbols;
        $rootScope.HeldSecurities = HeldSecurities;
        $rootScope.MarketStatus = MarketStatus;
        $rootScope.MySession = MySession;
        $rootScope.NewsSymbols = NewsSymbols;
        $rootScope.QuoteService = QuoteService;
        $rootScope.RecentSymbols = RecentSymbols;
        $rootScope.WebSockets = WebSockets;

    });

})();

File: app-js/src/main/scala/com/shocktrade/javascript/AppMain.scala
package com.shocktrade.javascript

import com.greencatsoft.angularjs.core._
import com.greencatsoft.angularjs.{Angular, Config}
import com.shocktrade.javascript.app.filter.YesNoFilter
import com.shocktrade.javascript.dashboard.{ContestServiceFactory, DashboardController, ExplorerServiceFactory}
import com.shocktrade.javascript.discover._
import com.shocktrade.javascript.news.{NewsServiceFactory, NewsSymbolsFactory}

import scala.language.experimental.macros
import scala.scalajs.js.JSApp
import scala.scalajs.js.annotation.JSExport

/**
 * ShockTrade Scala.js Application
 * @author lawrence.daniels@gmail.com
 */
@JSExport
object AppMain extends JSApp {

  /**
   * Angular.js application definition
   */
  override def main() {
    val app = Angular.module("shocktrade", Seq("ngAnimate", "ngCookies", "ngRoute", "ngSanitize", "nvd3ChartDirectives", "toaster", "ui.bootstrap"))
    app.config[RoutingConfig]

    // filters
    app.filter(YesNoFilter)

    // controllers
    app.controller[DashboardController]
    app.controller[DiscoverController]

    // services
    app.factory[ContestServiceFactory]
    app.factory[ExplorerServiceFactory]
    app.factory[FavoriteSymbolsFactory]
    app.factory[HeldSecuritiesFactory]
    app.factory[NewsServiceFactory]
    app.factory[NewsSymbolsFactory]
    app.factory[QuoteServiceFactory]
    app.factory[RecentSymbolsFactory]
  }

}

/**
 * Routing Config
 * @param routeProvider the given [[RouteProvider route provider]]
 */
@JSExport
class RoutingConfig(routeProvider: RouteProvider) extends Config {
  routeProvider
    .when("/connect", Route(templateUrl = "/assets/views/connect/connect.htm"))
    .when("/dashboard", Route(templateUrl = "/assets/views/dashboard/dashboard.htm"))
    .when("/dashboard/:contestId", Route(templateUrl = "/assets/views/dashboard/dashboard.htm"))
    .when("/discover", Route(templateUrl = "/assets/views/discover/discover.htm"))
    .when("/discover/:symbol", Route(templateUrl = "/assets/views/discover/discover.htm"))
    .when("/explore", Route(templateUrl = "/assets/views/explore/drill_down.htm"))
    .when("/inspect/:contestId", Route(templateUrl = "/assets/views/admin/inspect.htm"))
    .when("/news", Route(templateUrl = "/assets/views/news/news_center.htm"))
    .when("/research", Route(templateUrl = "/assets/views/research/research.htm"))
    .when("/search", Route(templateUrl = "/assets/views/play/search.htm"))
    .when("/symbols/favorites?:symbol", Route(templateUrl = "/assets/views/discover/favorites.htm" /* reloadOnSearch = false*/))
    .when("/symbols", Route("/symbols/favorites"))
    .when("/profile/awards", Route(templateUrl = "/assets/views/profile/awards.htm"))
    .when("/profile/statistics", Route(templateUrl = "/assets/views/profile/statistics.htm"))
    .when("/profile", Route("/profile/awards"))
    .otherwise(Route("/discover"))
}
